lib/oscap_tests.pm

```diff
@@ -0,0 +1,348 @@
+
+# Set default value for 'scap-security-guide' ds file
+our $f_ssg_sle_ds = '/usr/share/xml/scap/ssg/content/ssg-sle12-ds.xml';
+our $f_ssg_tw_ds = '/usr/share/xml/scap/ssg/content/ssg-opensuse-ds.xml';
+
+# Profile IDs
+# Priority High:
+our $profile_ID_sle_stig = 'xccdf_org.ssgproject.content_profile_stig';
+our $profile_ID_sle_cis = 'xccdf_org.ssgproject.content_profile_cis';
+our $profile_ID_sle_pci_dss = 'xccdf_org.ssgproject.content_profile_pci-dss';
+our $profile_ID_sle_hipaa = 'xccdf_org.ssgproject.content_profile_hipaa';
+our $profile_ID_sle_anssi_bp28_high = 'xccdf_org.ssgproject.content_profile_anssi_bp28_high';
+# Priority Medium:
+our $profile_ID_sle_anssi_bp28_enhanced = 'xccdf_org.ssgproject.content_profile_anssi_bp28_enhanced';
+our $profile_ID_sle_cis_server_l1 = 'xccdf_org.ssgproject.content_profile_cis_server_l1';
+our $profile_ID_sle_cis_workstation_l2 = 'xccdf_org.ssgproject.content_profile_cis_workstation_l2';
+# Priority Low:
+our $profile_ID_sle_anssi_bp28_intermediary = 'xccdf_org.ssgproject.content_profile_anssi_bp28_intermediary';
+our $profile_ID_sle_anssi_bp28_minimal = 'xccdf_org.ssgproject.content_profile_anssi_bp28_minimal';
+our $profile_ID_sle_cis_workstation_l1 = 'xccdf_org.ssgproject.content_profile_cis_workstation_l1';
+our $profile_ID_tw = 'xccdf_org.ssgproject.content_profile_standard';
+
+# The OS status of remediation: '0', not remediated; '1', remediated
+our $remediated = 0;
+
+# Upload HTML report by default
+set_var('UPLOAD_REPORT_HTML', 1);
+
+# Set value for 'scap-security-guide' ds file
+sub set_ds_file {
+
+    # Set the ds file for separate product, e.g.,
+    # for SLE15 the ds file is "ssg-sle15-ds.xml";
+    # for SLE12 the ds file is "ssg-sle12-ds.xml";
+    # for Tumbleweed the ds file is "ssg-opensuse-ds.xml"
+    my $version = get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
+    $f_ssg_sle_ds =
+      '/usr/share/xml/scap/ssg/content/ssg-sle' . "$version" . '-ds.xml';
+}
+
+sub upload_logs_reports {
+
+    # Upload logs & ouputs for reference
+    my $files;
+    if (is_sle) {
+        $files = script_output('ls | grep "^ssg-sle.*.xml"');
+    }
+    else {
+        $files = script_output('ls | grep "^ssg-opensuse.*.xml"');
+    }
+    foreach my $file (split("\n", $files)) {
+        upload_logs("$file");
+    }
+    upload_logs("$f_stdout") if script_run "! [[ -e $f_stdout ]]";
+    upload_logs("$f_stderr") if script_run "! [[ -e $f_stderr ]]";
+    if (get_var('UPLOAD_REPORT_HTML')) {
+        upload_logs("$f_report", timeout => 600)
+          if script_run "! [[ -e $f_report ]]";
+    }
+}
+
+sub pattern_count_in_file {
+
+    #Find count and rules names of matched pattern
+    my $self = $_[0];
+    my $data = $_[1];
+    my $pattern = $_[2];
+    my @rules;
+    my $count = 0;
+
+    my @lines = split /\n|\r/, $data;
+    for my $i (0 .. $#lines) {
+        if ($lines[$i] =~ /$pattern/) {
+            $count++;
+            push(@rules, $lines[$i - 4]);
```

Comment:
whant happens if $i < 4 ?

Reply:
Perl will fail, but he I rely on format of stdout:
Title
	Use Only FIPS 140-2 Validated MACs
Rule
	xccdf_org.ssgproject.content_rule_sshd_use_approved_macs_ordered_stig
Ident
	CCE-83280-8
Result
	pass
	
If [$i] exists -  [$i - 4 ] exists 100%

Reply:
the code relies on a specific output format, which can change and make it prone to unexpected behavior. Please consider adding at least a check on the array length and consider the usage of more idiomatic Perl constructs like foreach, grep. Perhaps you could also build a key,value hash with the data you expect instead of relying on fixed index offsets.

Reply:
> the code relies on a specific output format, which can change and make it prone to unexpected behavior.
We always rely on output format and cannot predict when and how it will change. There are no universal analysis method.  I will add additional checks to not get out of array.

>consider the usage of more idiomatic Perl constructs like foreach, grep.
Here **for my $i (0 .. $#lines)** equals **foreach** but with ability to get needed elements of array and easy rely on output format. Also in possible format change case can do easy fix. Grep function you referring does not fit the purpose.

>Perhaps you could also build a key,value hash with the data you expect instead of relying on fixed index offsets.
Do not see the point of building full key,value hash - I do not need all this data (Simpler is better and faster), only some of it.